File: 백준/Gold/15961. 회전 초밥/회전 초밥.py
# 회전 초밥
import sys
input = sys.stdin.readline

N, d, k, c = map(int,input().split())
sushi = []
types = [0] * (d+1)
cnt = 1

types[c] += 1

for i in range(N):
    sushi.append(int(input()))

for i in range(k):
    if types[sushi[i]] == 0:
        cnt += 1
        types[sushi[i]] = 1
    else:
        types[sushi[i]] += 1

result = cnt
for i in range(1, N):
    nex = sushi[(i + (k-1)) % N]
    prev = sushi[i-1]

    types[prev] -= 1
    if types[prev] == 0:
        cnt -= 1

    types[nex] += 1
    if types[nex] == 1:
        cnt += 1
        
    result = max(result, cnt)

print(result)   

File: 백준/Gold/15961. 회전 초밥/회전 초밥.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int d = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        int c = Integer.parseInt(st.nextToken());

        int[] sushi = new int[N];
        int[] types = new int[d+1];
        int cnt = 1;

        types[c]++;

        for (int i = 0; i < N; i++) {
            sushi[i] = Integer.parseInt(br.readLine());
        }

        for (int i = 0; i < k; i++) {
            if (types[sushi[i]] == 0) {
                cnt++;
                types[sushi[i]]++;
            } else {
                types[sushi[i]]++;
            }
        }

        int result = cnt;

        for (int i = 1; i < N; i++) {
            int next = sushi[(i + (k-1)) % N];
            int prev = sushi[i-1];

            types[prev]--;
            if (types[prev] == 0) {
                cnt--;
            }

            types[next]++;
            if (types[next] == 1) {
                cnt++;
            }

            result = Math.max(result, cnt);
            
        }

        System.out.println(result);
    }
}
